torch/csrc/jit/codegen/cuda/ir_utils.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/type.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+
+class Statement;
+
+namespace ir_utils {
+
+template <
+    typename FilterType,
+    template <typename, typename...> typename Container,
+    typename ElementType>
+Container<FilterType*> filterVals(const Container<ElementType*>& container) {
+  Container<FilterType*> out;
+  for (auto& s : container) {
+    if (s->getValType() == FilterType::type) {
```

Comment:
since you're interested in a better notation, would you be interested in adding another building block to the IR root class? ie. add this to Statement:

```cpp
 template<class T>
 bool IsA() const {
 return dynamic_cast<const T*>(this) != nullptr;
 }
```

this is a counterpart to Statement::as<T> that I have on my todo list, and would allow us to say:

`if (s->isA<FilterType>()) { ... }`

... and also eventually eliminate the need for ValType altogether.

Reply:
That looks nice, but one thing I didn't want to change is that the type checking is currently always done with `getValType()`. I think we can replace that with something like `IsA`, but I wonder whether `ValType` values and classes have one-to-one relationship. For example, would there be a case where there could be two different classes that are a subclass of `Val` AND have the same `ValType` value. I don't think that's the case now, but is this something that could happen in the future?

Reply:
I think switch to a more consistent `IsA` makes sense to me (uniformed API).
Can't we have specialization of `isA` so that we cannot necessarily do dynamic cast? that seems to also resolve the restriction on the one-to-one mapping.

```
 template<class T>
 bool IsA() const {
 return dynamic_cast<const T*>(this) != nullptr;
 }

 template<ValType T>
 bool IsA() const {
 return getValType() == T;
 }
```

Reply:
Using `IsA<>()` with the non-mapping filter and return would still require us to add an extra template parameter in the filterVals, (one for filter criterion used in the `IsA`, one to cast the return type `as`). Might bother people :shrug:

Reply:
@naoyam: a potential mismatch between the tag (`ValType`) and the actual type is one of the reasons I'm wary of using tags  , especially when they duplicate the type hierarchy 1:1 (like we do today). Since we're using a hierarchy to model the taxonomy of nodes, I think it's both easier and safer to use the types themselves. Another benefit of using the types (through a notation like `isA<T>`), is that you can query for abstract (non-leaf) types too, ex. `isA<Expr>` or `isA<Val>`

@jjsjann123 : as I explained above, the main benefit of `isA<T>` is not the notation itself, but the fact that it avoids tags altogether. So a version which tries to mix tags and types seems confusing at best and I'd recommend against it.

I'm not suggesting we should go as far as eliminating tags in this change. It seems an opportunity for adding `isA<T>` since it fits well with the proposed utility, but it's not really critical to this PR.

Reply:
I agree that we could get rid of the tag and just use the runtime type information would be more robust. However, I'd prefer to keep using the tag for type checking at least in this PR as changing that is not the purpose. I'll create a separate issue to consider.

Reply:
See #268